src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/PolymorphicTests.cs

```diff
@@ -529,6 +529,124 @@ public async Task AnonymousType()
             Assert.Equal(Expected, json);
         }
 
+        [Fact]
+        public async Task NestedPolymorphicClassesIncreaseReadAndWriteStackWhenNeeded()
+        {
+            // Regression test for https://github.com/dotnet/runtime/issues/71994
+            TestNode obj = new TestNodeList
+            {
+                Name = "testName",
+                Info = "1",
+                List = new List<TestNode>
+                {
+                    new TestLeaf
+                    {
+                        Name = "testName2"
+                    },
+
+                    new TestNodeList
+                    {
+                        Name = "testName3",
+                        Info = "2",
+                        List = new List<TestNode>
+                        {
+                            new TestNodeList
+                            {
+                                Name = "testName4",
+                                Info = "1"
+                            }
+                        }
+
+                    }
+                }
+            };
+
+            string json = await Serializer.SerializeWrapper(obj);
+            JsonTestHelper.AssertJsonEqual(
+                @"{
+                   ""$type"": ""NodeList"",
+                   ""Info"": ""1"",
+                   ""List"": [
+                     {
+                       ""$type"": ""Leaf"",
+                       ""Test"": null,
+                       ""Name"": ""testName2""
+                     },
+                     {
+                       ""$type"": ""NodeList"",
+                       ""Info"": ""2"",
+                       ""List"": [
+                         {
+                           ""$type"": ""NodeList"",
+                           ""Info"": ""1"",
+                           ""List"": null,
+                           ""Name"": ""testName4""
+                         }
+                       ],
+                       ""Name"": ""testName3""
+                     }
+                   ],
+                   ""Name"": ""testName""}", json);
+
+            TestNode deserialized = await Serializer.DeserializeWrapper<TestNode>(json);
+            obj.AssertEqualTo(deserialized);
+        }
+
+        [JsonDerivedType(typeof(TestNodeList), "NodeList")]
```

Comment:
I had added a test for recursive polymorphic types, but it seems to have stopped short at depth == 3:

https://github.com/dotnet/runtime/blob/87aa786b3997a23501912abd4654e4fd9958230c/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/PolymorphicTests.CustomTypeHierarchies.cs#L1744-L1753

Could you add a test that checks `Peano.FromInteger(size)` with large enough sizes (i.e. 10 or 64)? You can skip handcrafting the expected JSON and just do a roundtrip test (you might need an additional `Peano.ToInteger` implementation to validate the deserialized output).

Reply:
For context Peano numbers are an encoding for natural numbers using linked lists. The inverse operation could be implemented as follows:
```C#
public static int ToInteger(Peano value) => value is Succ succ ? 1 + ToInteger(succ.value) : 0; 
```